### ISSUE TITLE
New method Interests (Not 100% correct)

### ISSUE THREAD

**Issue by GeorgeFive** [duck7000/imdbGraphQLPHP#73]
Hey, I noticed imdb changed genres to interests, and there's now a lot more detail there.

https://www.imdb.com/title/tt0077402/
https://www.imdb.com/interest/in0000122/

This is actually pretty cool, and is pretty handy info to have.  Can we get this over here?

**Comment by duck7000** [comment]
The new style looks a lot like netflix hah
But yes this looks interesting!

I'm not sure how this can be used in our library?
With new methods i suppose?
Or something like Calendar class?

I need some examples or ideas how to use this

**Comment by GeorgeFive** [comment]
I just looked at this again on mobile, it's different now. Earlier, the genres on the main title page were gone and replaced with interests.. Maybe I'm on the test site at home? Who knows. 

But mainly, I'd just like to get a list of the interests in the way that genres currently works.

Not sure if the old genre links will still work when this rollout is complete.....?

**Comment by duck7000** [comment]
I'm not sure either..
It can take a lot of time for new stuff to rollout to all imdb servers so we may have to wait a while before trying this.

So a new method interests in title class would be what you want?
It might be better to do this in a similar class like Calendar i think?

First i have to check if and how this is implanted in GraphQL API

I'll look in to this but first i want to sort of close the musicBrainz library with a new version. I can't work on both at the same time

**Comment by GeorgeFive** [comment]
![Clipboard01](https://github.com/user-attachments/assets/20ec9061-d181-4269-8a36-43a6677f0684)

Ok, so I didn't see this on mobile at work, but now that I'm home, this is what I'm seeing (just in case you don't have it on your end yet).

I think it would be nice to have it in title, just get the list of interests and ids from a given title (same thing as current genres method).  That's all I personally need, but if there was a separate class to get everything from an interest id, that would be cool as well.

Take your time, as always!  No rush on this, I definitely don't need it today, hah.  I'm wrapping up the musicbrainz stuff as well, getting everything done on my end to actually use the class (hence the rapid fire reports on here, haha).

**Comment by duck7000** [comment]
Well i looked in to this and found that imdb now uses genres AND subgenres by keyword

In the genre method i did only include genre (not sure if subGenre existed at that time).
So i'm going to include subGenre as well but have to see how this would be added in the return results?

**Comment by GeorgeFive** [comment]
So they return as, ie, "Zombie Horror" is a subgenre of "Horror"?  And then it has horror as well?

Maybe we could have a header_id on subgenres to tie everything together?

**Comment by duck7000** [comment]
Yes as far as i have seen they use genre and subGenres

Mm i'm thinking of keep all genres together in an indexed array just as it is now, not sure if i can find header_id though

**Comment by GeorgeFive** [comment]
I'm not sure if they actually call it header_id or not, just something to say that X is a subgenre of Y.

My thought process is that, as a human, of course you could tell that "Zombie Horror" is a subgenre of "Horror", that's simple.... but how do you determine that with code?  You could match against "Horror", but what if they have a genre called "Vampires"?

If it's buried in there, it would definitely be cool to have!

**Comment by duck7000** [comment]
The API keeps genre and subGenre separated so i can add them separated to the output array?
a associative array: 
array(['mainGenre']) where mainGenre is an array with subGenres?

**Comment by GeorgeFive** [comment]
Gotcha, that could work!  Just tossing out ideas, if it's possible, awesome... if not, I'll see what I can do, haha.

**Comment by duck7000** [comment]
subGenre is added in latest commit.
The output array is changed, see here

```
Array
(
    [0] => Array
        (
            [mainGenre] => Crime
            [subGenre] => Array
                (
                )

        )

    [1] => Array
        (
            [mainGenre] => Sci-Fi
            [subGenre] => Array
                (
                    [0] => dystopian sci fi
                )

        )

)
```

Let me know if this works for you?

**Comment by duck7000** [comment]
The other part about interests i have to try to find it in the API, i didn't find it jet

**Comment by GeorgeFive** [comment]
Just tried it out... I noticed that not everything is returning?

https://www.imdb.com/title/tt0077402/

If you look at the site, they have a "Survival" genre as well, but this is not returned.  Also, do they provide the genre ID's by any chance?

```

Array
(
    [0] => Array
        (
            [mainGenre] => Horror
            [subGenre] => Array
                (
                    [0] => zombie horror
                    [1] => supernatural horror
                )

        )

    [1] => Array
        (
            [mainGenre] => Thriller
            [subGenre] => Array
                (
                )

        )

)
```

**Comment by duck7000** [comment]
well for dawn of the dead this is returned

```
stdClass Object
(
    [title] => stdClass Object
        (
            [titleGenres] => stdClass Object
                (
                    [genres] => Array
                        (
                            [0] => stdClass Object
                                (
                                    [genre] => stdClass Object
                                        (
                                            [text] => Horror
                                        )

                                    [subGenres] => Array
                                        (
                                            [0] => stdClass Object
                                                (
                                                    [keyword] => stdClass Object
                                                        (
                                                            [text] => stdClass Object
                                                                (
                                                                    [text] => zombie horror
                                                                )

                                                        )

                                                )

                                            [1] => stdClass Object
                                                (
                                                    [keyword] => stdClass Object
                                                        (
                                                            [text] => stdClass Object
                                                                (
                                                                    [text] => supernatural horror
                                                                )

                                                        )

                                                )

                                        )

                                )

                            [1] => stdClass Object
                                (
                                    [genre] => stdClass Object
                                        (
                                            [text] => Thriller
                                        )

                                    [subGenres] => Array
                                        (
                                        )

                                )

                        )

                )

        )

)
```

I can't produce what is not returned, but yes there is difference between what is shown above under the trailer and what is shown by genres (if you scroll down the page)

This is it i'm afraid

**Comment by duck7000** [comment]
Genre id's i'll have to look in to that

**Comment by duck7000** [comment]
Mm the genres on top of the page are interest, further down are genres

This is confusing!

If i search in API with the word 'interest' there is al whole list but nothing that indicates something about genres

**Comment by GeorgeFive** [comment]
If I had to guess, I'd say that interests will replace genres soon.  They just launched interests a few nights ago, so I think it's still a little rough, but give it a few weeks, hah.

Yeah, I never used genres before this, because "Horror" "Thriller" is a little vague for my liking (we do our own genres on my site to really categorize titles).  But this interests thing is pretty nice, I do have to admit that.  That's why I got "interest"ed all of a sudden, hah.

**Comment by duck7000** [comment]
That might indeed happen but we'll see
Her is some info
https://help.imdb.com/article/issues/GG7J2ARMQ6SFY7AF

There is no API documentation and no release notes either so for now this is the best i can do

If i do find something useful i let you know

**Comment by GeorgeFive** [comment]
Any updates on this?  genre() is pretty close, but it doesn't include everything.  Same issue as up above there...

Example: https://www.imdb.com/title/tt0077402/

Array
(
    [0] => Array
        (
            [mainGenre] => Horror
            [subGenre] => Array
                (
                    [0] => Zombie Horror
                    [1] => Supernatural Horror
                )
        )
    [1] => Array
        (
            [mainGenre] => Thriller
            [subGenre] => Array
                (
                )
        )
)

"Survival" isn't included in this list.  Also, if possible, I'd still love to have the genre ID in there!

**Comment by duck7000** [comment]
First there is no list of interests or interests Id, there is no metadata for it.
I'm making a list of interests according to the interests page but it is slow progress as it all is manual work.

And Survival is NOT in GraphQL genre list of this title, so i can't add what is not there.
While there is a genreId this is in text form, not in id form. You might expect this would be a "gr00012568" or "ge125898" or something but it is not.
GenreId is just like the name of genre, example: genre name "Horror" would have id "Horror"
This might be the reason they are calling it interests now as interests DO have a id like "in000115"

So my hands are tight and yes they might switch to interests instead of genre but there is no indication if or when anywhere.

Through mostPopularTitle() in Chart class you can recreate the interests pages but this needs a list of genres as well as subGenres, so it needs the list to be complete, and the results are max 30 or so.

```
Action in0000001
    Action Epic             in0000002
    B-Action                in0000003
    Car Action              in0000004
    Disaster                in0000005
    Gun Fu                  in0000197
    Kung Fu                 in0000198
    Martial Arts            in0000006
    One-Person Army Action  in0000007
    Samurai                 in0000199
    Superhero               in0000008
    Sword & Sandal          in0000009
    War                     in0000010
    War Epic                in0000011
    Wuxia                   in0000200
```
So far i have a start of that list, as you see this is slow progress. And there seems to be some kind of structure id in it as easter interest (kung fu, samurai and wuxia) have a number that start with 100

So no there is no practical progress but behind the scenes i'm trying but it seems a uphill battle so far

**Comment by GeorgeFive** [comment]
Gotcha.  Yeah, I do remember all these issues from before, but I was curious to see if anything had been updated on their end.  I understand that if they don't provide the data, you can't get the data, so we're essentially stuck.  It's been four months, so I thought maybe they would have done something with this!?  I guess not.

**Comment by duck7000** [comment]
progress is slow at imdb, they are still converting some parts of their website to the new layout/style, a page could take more then a year and still credit page from a title is not converted jet.

So they introduce new things but the progress with existing changes is slow

so yes... where stuck for the moment, and that might be so for a long time as i wouldn't expect any changes within a year
If or when they switch to interests instead of genres remains a question although i would agree with you that they eventually will

**Comment by GeorgeFive** [comment]
Six months later... any updates on their end?  Same as before, I understand you can't get the data if they don't provide it, just checking to see if they've made progress, hah.

**Comment by GeorgeFive** [comment]
hmmm.... so I just looked into this a bit, and you're right, these are keywords. If you look at they keywords for a title, they now have a category attached. If category=subgenre, it shows on the front page.

Plot Detail army
Plot Detail army-special-operation
Plot Detail attack
Plot Detail axe
Subgenre b-horror
Plot Detail back-from-the-dead
Plot Detail bad-guy
Plot Detail bald-zombie

The current keyword() function does not return category.  Is it possible to grab that info?  If so, we could also add this to genres....?

**Comment by duck7000** [comment]
Mm keyword category can be added to keyword(), as text or as categoryId (Plot Details vs plot-detail)

I am not sure how to use/add this to genres?

**Comment by duck7000** [comment]
Currently the query from genre() already uses keyword

```
query Genres(\$id: ID!) {
  title(id: \$id) {
    titleGenres {
      genres {
        genre {
          text
        }
        subGenres {
          keyword {
            text {
              text
            }
          }
        }
      }
    }
  }
}
EOF;
```
That is where the subGenre is coming from

**Comment by duck7000** [comment]
Mm i am not against the idea to add a genreKeywords() method but i have to think about how to accomplish combining them together.

And at the top of the title page you mentioned Dark Comedy is listed (as interest) but this is not a genre but a keyword. So they mix up keywords and genres but also use keywords as interest/genre

So in short interests consist of multiple parts like genre and keywords in any form


This is a thought about if i can get interests info through titleGenre in GraphQL:
I have checked genre in GraphQL and there is no output (for this title) Dark comedy or Survival. 
Those words are independently there but not in combination with each other.
e.g Dark is there (as dark fantasy) or  Survival (as survival adventure)

But this info is coming as sub divisions from titlegenres/subgenres/keywords/
In short: this info are all titles related of subgenres keywords from genres (i hope you can follow it)

```
query {
  title(id:"tt0077402") {
    titleGenres {
      genres {
        subGenres {
          keyword {
            titles(first:9999){
              edges {
                node {
                  titleGenres {
                    genres {
                      subGenres {
                        keyword {
                          text {
                            text
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}
```
This way provides more info/genres/keywords but it makes it very complex to process as they appear the same multiple times.
And there is no way to tell how imdb decides to use the info as interests though?
So you see "it's complicated"

**Comment by duck7000** [comment]
This is a small part of the source code of the title page of Dawn of the dead. It is coming from a JavaScript part.

`"interests":{"edges":[{"node":{"id":"in0000108","primaryText":{"text":"B-Horror","__typename":"InterestText"},"__typename":"Interest"},"__typename":"InterestEdge"},{"node":{"id":"in0000035","primaryText":{"text":"Dark Comedy","__typename":"InterestText"}`

Dark comedy is coming from interests, InterestText,primaryText

So you could think it is possible to get..Nope!
Interests is not in GraphQL (jet?) so for now this is still a dead end i'm afraid.
It is not available in the public GraphQL API, when/if is unsure at this time

In my above examples/code it is impossible to know what those interests are

**Comment by duck7000** [comment]
Well to get those exact interests, no not possible. Imdb might add that to the public GraphQL API but unsure at this moment.

A new method keywordGenres() might be a little in the right direction but still it is impossible to know which exact keyword imdb chooses to use as interest. So it could return all keywords with category subGenre but that still is not accurate as it will return possibly hundreds of keywords

So the only thing that can work in the right direction is adding category's to keyword() and leave it up to the user how to work with that. This will require a change to the output array.

So in short, there is no exact solution

**Comment by GeorgeFive** [comment]
At least with the original example (again, haven't tested loads of movies), they only have 7 keywords labeled as subgenres, and 6 of those show on the front page ("satire comedy" is the only one that doesn't display - maybe there's a mainpage limit).  This is out of the 844 keywords for this title, so I definitely think this is the key, as aside from the base genres (horror / thriller), everything from the main page is there in keywords with the subgenre category, with all but 1 being used (and no false positives).

I think the new method is a good idea for completion, but if we can get category for keywords, I could achieve the same thing in my own code.

---

Go to edit page at the bottom, keywords, correct/delete... there's the list with categories.

**Comment by GeorgeFive** [comment]
Sounds like a plan!  Maybe we can finally put this issue to bed, haha.

**Comment by duck7000** [comment]
I made some progress

```
Array
(
    [0] => dark comedy
    [1] => zombie horror
    [2] => supernatural horror
    [3] => satire comedy
    [4] => survival adventure
    [5] => b horror
    [6] => splatter horror
)
/

Array
(
    [0] => Array
        (
            [mainGenre] => Horror
            [subGenre] => Array
                (
                    [0] => Zombie Horror
                    [1] => Supernatural Horror
                    [2] => B Horror
                    [3] => Splatter Horror
                )

        )

    [1] => Array
        (
            [mainGenre] => Thriller
            [subGenre] => Array
                (
                )

        )

)
/
```

The first array are keywords with categoryId subgenre, so far so good but notice that there is one extra, satire comedy, and survival is survival adventure.
Next it is missing 2 main genres: Horror and Thriller.

The missing main genres (the second array) i can get from genre() and add them to interests()
So it is not 100% accurate but the best i can do/get.

**Comment by duck7000** [comment]
Nope, Interest id is not possible, as this is a keyword not an interest.
I can add keywordId for example: kw0035381 but that doesn't help i guess?

**Comment by GeorgeFive** [comment]
Gotcha! I can definitely work something out on my end, maybe just attach a number to each unique string and use that.

I think it's ready to go!

**Comment by duck7000** [comment]
So my latest suggesting is also not watertight..

Remains starting at 1 and counting on like this 'in0000001'?
Or only number 1 and counting on
This system is unique within a title, but the same for every title

**Comment by GeorgeFive** [comment]
It might be best to just skip the id if we can't get legit info from them. I can see confusion from users if we say zombie horror is id 1234, but on IMDb, zombie horror is actually id 5678. If we can't grab that legitimate 5678 number, we should probably not return anything. 

However, if we're determined to get a unique id, maybe the PHP function crc33 would help?

**Comment by GeorgeFive** [comment]
Looks perfect, have tested a lot of movies and seeing no issues. Thanks!

**Comment by duck7000** [comment]
Thanks, and your welcome

**Comment by GeorgeFive** [comment]
Yeah, looks like B Western is a keyword.  Here's one that is tagged with that: https://www.imdb.com/title/tt0190229

Let's see what I can do when this finishes scanning and I get out of work.... can at least get a majority.

**Comment by GeorgeFive** [comment]
Keep in mind that your interests list there and the list we're currently getting is slightly different in some ways 

Example, I see you're getting Kaiju, but the script is returning Giant Monster Sci Fi..... you're getting Korean Drama, but the script returns K Drama..... etc.

**Comment by GeorgeFive** [comment]
Just thinking out loud here. If we get the interest as is, ie, K Drama.... and then try to match it to an ID from your list.... we're going to need an alias setup or something.

Or just ignore those names and do some internal logic? We know K Drama = Korean Drama, so use that id if the function returns K Drama.

**Comment by duck7000** [comment]
Maybe we can manually can go through both lists and make one correct list of it?
If we can find the differences between those lists it is possible i guess.
And 200 items is doable by hand

**Comment by GeorgeFive** [comment]
Ok, so that's a keyword.

https://www.imdb.com/title/tt26315487

Which is weird, because this title has Korean Drama as a legitimate interest, but we don't get that... but we do get K Drama, which has the subgenre category.  So I'm thinking it's the same thing in the background, but with different display names....???  Just a guess at this point.

**Comment by duck7000** [comment]
Mm this is in my scraped array: 'Korean Drama' => 'in0000209', so not a keyword i guess?

But yes this are the "flaws" in our interests() method..
That is the difference between our "interests()" method based on keywords versus imdb "interests"
imdb uses K Drama from keywords but converts that to Korean Drama as interest

We could add this kind of flaws to my scraped array, so in this case there will be 2 entry's with the same id:
K Drama => in0000209
Korean Drama => in0000209

This way we can get the right id but it returns K Drama as interest

**Comment by GeorgeFive** [comment]
Yeah, I think we're good on this one, thanks!
